daemon: start DNS proxy before k8s

jrajahalme opened the pull request:
Split the endpoint restoration process into more stages so that the saved endpoint DNS history can be used earlier in the bootstrap to start the DNS proxy. This helps ensure that the first regeneration of an Endpoint does not (temporarily) remove the policy keys needed to keep old traffic flowing.

TPROXY rules for the DNS proxy can only be installed couple of seconds later in the bootstrap process. To mitigate this and allow the DNS proxy start serving traffic sooner, it now reads the old TPROXY rule, if available, and tries to re-use the same proxy port number as before the restart. This allows the old datapath config to start feeding traffic to the new proxy instance even before the datapath config can be rewritten.

For new DNS resolutions to be added to the Endpoint's policy the restored endpoint's must still have their first regeneration completed. Before this it is also likely that the DNS proxy will not serve requests but answers with an error response instead.

Fixes: #11004

```release-note
DNS Proxy is started earlier in the Cilium agent bootstrap to make it available to running endpoints sooner.
```

tklauser commented on daemon/cmd/daemon.go:
```diff
@@ -404,6 +404,45 @@ func NewDaemon(ctx context.Context, epMgr *endpointmanager.EndpointManager, dp d
+	}
+
+	bootstrapStats.restore.Start()
+	// fetch old endpoints before k8s is condifured.
```
Nit:
```suggestion
	// fetch old endpoints before k8s is configured.
```

jrajahalme replied:
fixed.

tklauser commented on daemon/cmd/fqdn.go:
```diff
@@ -263,20 +263,20 @@ func (d *Daemon) bootstrapFQDN(restoredEndpoints *endpointRestoreState, preCache
 	globalCache := d.dnsNameManager.GetDNSCache()
 	now := time.Now()
-	for _, restoredEP := range restoredEndpoints.restored {
+	for _, possibleEP := range possibleEndpoints {
```
Making `possibleEndpoints` a map (I assume due to the return type of `ReadEPsFromDirNames`?) will lead to random iteration order (see https://blog.golang.org/maps, section _Iteration order_). Could this become a problem?

jrajahalme replied:
Good point, but that should not make any difference here.

tklauser commented on daemon/cmd/state.go:
```diff
@@ -90,6 +91,39 @@ func (d *Daemon) validateEndpoint(ep *endpoint.Endpoint) (valid bool, err error)
+	}
+	eptsID := endpoint.FilterEPDir(dirFiles)
+
+	state.possible = endpoint.ReadEPsFromDirNames(d.ctx, d, dir, eptsID)
```
Might be worth to change `possible` to be `[]*endpoint.Endpoint` and convert the return value of `endpoint.ReadEPsFromDirNames` to it in order to avoid the random iteration order issue pointed out above, e.g.:

```suggestion
	possible = endpoint.ReadEPsFromDirNames(d.ctx, d, dir, eptsID)
	state.possible = make([]*endpoint.Endpoint, 0, len(possible)
	for _, p := range possible {
		state.possible = append(state.possible, p)
	}
	sort.Slice(state.possible, func(i, j int) bool {
		return state.possible[i].ID < state.possible[j].ID
	}
```

jrajahalme replied:
As said above, the iteration order should make no difference.

jrajahalme committed:
daemon: Minimize DNS proxy donwtime on restart

Start proxy support earlier in the daemon bootstrap, notably before any k8s setup.

Fetch old endpoints earlier so that the DNS history is avalailable
before k8s is set up and move dns proxy initialization earlier in the
bootstrap.

Reuse DNS proxy port from previous run on restart unless overridden by
an explicit Cilium agent option

These changes allow the DNS proxy start serving requests as soon as
the toFQDN policy is received from k8s and avoid any service
disruption prevously possible due to endpoints being regenerated
before the DNS proxy was started.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

tklauser reviewed (commented):
> [...] and tries to re-use the same proxy port number as before the restart.

Given the port reuse, should we set the `SO_REUSEPORT` sockopt on the proxy socket to avoid potential problems when re-`bind()`ing on agent restarts?
